File: actojat-c-support/src/main/java/de/netherspace/apps/actojat/languages/c/CVisitor.java
package de.netherspace.apps.actojat.languages.c;

import de.netherspace.apps.actojat.c_grammarBaseVisitor;
import de.netherspace.apps.actojat.c_grammarParser;
import de.netherspace.apps.actojat.ir.java.ArgumentDeclaration;
import de.netherspace.apps.actojat.ir.java.Assignment;
import de.netherspace.apps.actojat.ir.java.Expression;
import de.netherspace.apps.actojat.ir.java.ForLoop;
import de.netherspace.apps.actojat.ir.java.FunctionCall;
import de.netherspace.apps.actojat.ir.java.Import;
import de.netherspace.apps.actojat.ir.java.JavaLanguageConstruct;
import de.netherspace.apps.actojat.ir.java.LeftHandSide;
import de.netherspace.apps.actojat.ir.java.Method;
import de.netherspace.apps.actojat.ir.java.Program;
import de.netherspace.apps.actojat.ir.java.Statement;
import de.netherspace.apps.actojat.languages.BaseVisitor;
import kotlin.Pair;
import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;

import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

/**
 * A visitor implementation that generates an intermediate representation for a
 * particular parse tree.
 */
public class CVisitor extends c_grammarBaseVisitor<JavaLanguageConstruct> implements BaseVisitor {

  private final Map<String, Method> methods = new HashMap<>();
  private final List<Import> imports = new LinkedList<>();

  @Override
  public JavaLanguageConstruct visit(ParseTree tree) {
    super.visit(tree);
    return new Program(methods, imports, null);
  }


  /**
   * Maps a list of C arguments to a HashMap.
   */
  Function<c_grammarParser.ArgumentlistContext, Map<String, String>> argumentsToJavaArgs = args -> {
    //TODO!
    return null;
  };


  /**
   * Maps a C assignment operator to a Java assignm. op.
   */
  Function<TerminalNode, String> assignmentOperatorToJavaOperator = op -> {
    return op.getText();
  };


  @Override
  public JavaLanguageConstruct visitFunctiondeclr(c_grammarParser.FunctiondeclrContext ctx) {
    final String methodName = computeMethodName(ctx);
    final String sourceName = methodName; // TODO: this doesn't look right... Fix it!

    /*List<Argument> jarguments = argumentsToJavaArgs.apply(ctx.argumentlist())
                             .entrySet()
                             .stream()
                             .map(argEntryToJavaArgument)
                             .collect(Collectors.toList());
    */
    final List<ArgumentDeclaration> arguments = Collections.emptyList(); // TODO: fix the above code!

    final List<Statement> statements = ctx.block()
        .expressionlist()
        .expression()
        .stream()
        .map(expressionToJavaStatement)
        .collect(Collectors.toList());

    final Method javaMethod = new Method(methodName, statements, arguments, null);
    methods.put(sourceName, javaMethod);
    return javaMethod;
  }


  /**
   * Maps a C function to a Java method's name.
   */
  private String computeMethodName(c_grammarParser.FunctiondeclrContext ctx) {
    return ctx.ID().getText();
  }


  /**
   * Maps a C parameter (e.g. the "HelloWorld" in
   * printf("HelloWorld")
   * ) to a Java expression.
   */
  private final Function<c_grammarParser.ArgumentContext, Expression> parameterToJavaExpression
      = param -> {
        final String[] parts = {param.getText()};
        return new Expression(parts, null);
      };


  /**
   * Maps a C expression to a Java statement.
   */
  private Function<c_grammarParser.ExpressionContext, Statement> expressionToJavaStatement = ex -> {
    // TODO: distinguish whether its a function call via the grammar!
    if (ex.functioncall() != null) {
      // set the function's name (e.g. 'doSomething' for 'bla = doSomething();' ):
      final String functionName = ex.functioncall().ID().getText();

      if (ex.functioncall().argumentlist() != null) {
        final List<Expression> parameters = ex.functioncall()
            .argumentlist()
            .argument()
            .stream()
            .map(this.parameterToJavaExpression)
            .collect(Collectors.toList());
        return new FunctionCall(functionName, parameters, null);
      } else {
        final List<Expression> parameters = Collections.emptyList();
        return new FunctionCall(functionName, parameters, null);
      }
    }

    // its a mere assignment:
    if (ex.assignment() != null) {
      return this.assignmentToJavaAssignment.apply(ex.assignment());
    }

    if (ex.returnstatement() != null) {
      final String functionName = "return"; // TODO: create an enum holding these values!
      final List<Expression> parameters = Collections.emptyList();
      final FunctionCall functionCall = new FunctionCall(functionName, parameters, null);
      // ...
      return functionCall;
    }

    if (ex.forloop() != null) {
      final c_grammarParser.ForloopContext forLoop = ex.forloop();
      return this.forLoopToJavaForLoop.apply(forLoop);
    }

    System.err.println("couldn't determine statement type:");
    System.err.println(" " + ex.getText());
    return null;
  };


  /**
   * Maps a C for-loop to a Java for-loop.
   */
  private Function<c_grammarParser.ForloopContext, ForLoop> forLoopToJavaForLoop = fl -> {
    Assignment loopVariable = null;
    if (fl.assignment() != null) {
      // Case 1: "for (int i=0, ..."
      loopVariable = this.assignmentToJavaAssignment.apply(fl.assignment());

    } else if (fl.rhs() != null) {
      // Case 2: "for (i=0, ..."
      final String rhs = fl.rhs().getText();
      //loopVariable.setRhs(rhs); // TODO: does it work just like "fl.assignment()"?
      loopVariable = this.assignmentToJavaAssignment.apply(fl.assignment());

    } else {
      System.err.println("No loop variable present!");
      return null;
    }

    String loopCondition = fl.condition().getText();
    String loopIncrement = fl.incrementstatement().getText();

    final Statement[] body = fl
        .block()
        .expressionlist()
        .expression()
        .stream()
        .map(expressionToJavaStatement)
        .toArray(Statement[]::new);

    return new ForLoop(loopVariable, loopCondition, loopIncrement, body, null);
  };


  /**
   * Maps a C assignment to a Java assignment.
   */
  private Function<c_grammarParser.AssignmentContext, Assignment> assignmentToJavaAssignment
      = assignment -> {
        final Pair<String, String> lhs = this.computeLeftHandSide(assignment.lhs());
        final String lhsTypeAnnotation = lhs.getFirst();
        final String lhsVariableName = lhs.getSecond();
        final LeftHandSide jlhs = new LeftHandSide(lhsTypeAnnotation, lhsVariableName);
        final String jrhs = this.computeRightHandSide(assignment.rhs());
        return new Assignment(jlhs, jrhs, null);
      };


  /**
   * Maps a left-hand side identifier to a Java identifier: type x name.
   */
  private Pair<String, String> computeLeftHandSide(c_grammarParser.LhsContext lhs) {
    if (lhs.variabledecl() != null) {
      final String typeAnnotation = lhs.variabledecl().primitivetype().getText();
      final String variableName = lhs.variabledecl().ID().getText();
      return new Pair<>(typeAnnotation, variableName);

    } else {
      final String variableName = lhs.ID().getText();
      return new Pair<>(null, variableName);
    }
  }


  /**
   * Maps a right-hand side to a Java RHS.
   */
  private String computeRightHandSide(c_grammarParser.RhsContext rhs) {
    return rhs.getText(); // TODO: recursively handle nested expressions!
  }


  @Override
  public JavaLanguageConstruct visitImportheader(c_grammarParser.ImportheaderContext ctx) {
    final String importName = computeImportName(ctx);
    final Import jimport = new Import(importName, null);
    imports.add(jimport);
    return jimport;
  }


  /**
   * Maps a C include to a Java's import file name.
   */
  private String computeImportName(c_grammarParser.ImportheaderContext include) {
    final String includeWithoutSeparatingDot = include.FILEID()
        .getText().replaceAll("\\.", "_");
    final String includeWithoutSlashes = includeWithoutSeparatingDot
        .replaceAll("/", "_");
    final String includeWithoutDashes = includeWithoutSlashes
        .replaceAll("-", "_");
    return includeWithoutDashes;
    // TODO: blacklist standard libs (e.g. "<stdio.h>")!
  }

}

File: actojat-c-support/src/main/kotlin/de/netherspace/apps/actojat/languages/c/CVisitor.kt
package de.netherspace.apps.actojat.languages.c

import de.netherspace.apps.actojat.c_grammarBaseVisitor
import de.netherspace.apps.actojat.c_grammarParser
import de.netherspace.apps.actojat.ir.java.*
import de.netherspace.apps.actojat.languages.BaseVisitor
import de.netherspace.apps.actojat.languages.JavaIrUtil
import org.antlr.v4.runtime.tree.ParseTree
import org.slf4j.LoggerFactory

class CVisitor : c_grammarBaseVisitor<List<JavaLanguageConstruct>>(), BaseVisitor {

    private val log = LoggerFactory.getLogger(CVisitor::class.java)

    private var className: String? = null
    private val methods = mutableMapOf<String, Method>()
    private val imports = mutableListOf<Import>()

    override fun visit(tree: ParseTree?): List<JavaLanguageConstruct> {
        super.visit(tree)
        return listOf(Clazz(
                className = className,
                methods = methods,
                imports = imports,
                fields = mapOf(),
                comment = null
        ))
    }

    override fun visitFunctiondeclr(ctx: c_grammarParser.FunctiondeclrContext?): List<JavaLanguageConstruct> {
        val methodName: String = computeMethodName(
                ctx ?: throw NullPointerException("Got a null value from the AST")
        )
        val sourceName = methodName // TODO: this doesn't look right... Fix it!

        /*List<Argument> jarguments = argumentsToJavaArgs.apply(ctx.argumentlist())
                             .entrySet()
                             .stream()
                             .map(argEntryToJavaArgument)
                             .collect(Collectors.toList());
        */
        val arguments: List<ArgumentDeclaration> = listOf()// TODO: fix the above code!

        val statements: List<Statement> = expressionListToJavaStatements(ctx.block().statementlist())

        val javaMethod = Method(
                name = methodName,
                statements = statements,
                arguments = arguments,
                comment = null
        )
        methods[sourceName] = javaMethod
        return listOf(javaMethod)
    }

    override fun visitImportheader(ctx: c_grammarParser.ImportheaderContext?): List<JavaLanguageConstruct>? {
        // blacklist standard libs (e.g. "<stdio.h>"):
        val importBlacklist = listOf("stdio.h")
        val rawImportName: String = ctx?.FILEID()?.text
                ?: throw NullPointerException("Got a null value from the AST")
        return if (!importBlacklist.contains(rawImportName)) {
            val importName: String = computeImportName(ctx)
            val jimport = Import(
                    name = importName,
                    comment = null
            )
            imports.add(jimport)
            return listOf(jimport)
        } else {
            listOf()
        }
    }

    /**
     * Maps a C function to a Java method's name.
     */
    private fun computeMethodName(ctx: c_grammarParser.FunctiondeclrContext): String {
        return ctx.ID().text
    }

    /**
     * Maps a C expression to a Java statement.
     */
    private fun expressionToJavaStatement(ex: c_grammarParser.StatementContext?): Statement {
        if (ex == null) {
            throw NullPointerException("Got a null value from the AST")
        }

        // TODO: distinguish whether its a function call via the grammar!
        // TODO: use a "when"!

        // is it a function call?
        if (ex.functioncall() != null) {
            // set the function's name (e.g. 'doSomething' for 'bla = doSomething();' ):
            val functionName = ex
                    .functioncall()
                    ?.ID()
                    ?.text
                    ?: throw NullPointerException("Got a null value from the AST")

            val parameters = if (ex.functioncall()?.argumentlist() != null) {
                ex.functioncall()
                        .argumentlist()
                        .argument()
                        .map { parameterToJavaExpression(it) }
                        .toList()
            } else {
                listOf()
            }
            return FunctionCall(
                    name = functionName,
                    parameters = parameters,
                    comment = null
            )
        }

        // it's a mere assignment:
        if (ex.assignment() != null) {
            return assignmentToJavaAssignment(ex.assignment())
        }

        // it's a "return":
        if (ex.returnstatement() != null) {
            val functionName = "return" // TODO: create an enum holding these values!
            return FunctionCall(
                    name = functionName,
                    parameters = listOf(),
                    comment = null
            )
        }

        // it's a for-loop:
        if (ex.forloop() != null) {
            return forLoopToJavaForLoop(ex.forloop())
        }

        // it's an if-then-else conditional expr.:
        if (ex.ifthenelse() != null) {
            return ifthenelseToJavaConditionalExpr(ex.ifthenelse())
        }

        throw Exception("couldn't determine statement type:" + ex.text)
    }

    /**
     * Maps a C parameter (e.g. the "HelloWorld" in
     * printf("HelloWorld")
     * ) to a Java expression.
     */
    private fun parameterToJavaExpression(param: c_grammarParser.ArgumentContext?): Expression {
        val parts = arrayOf(param?.text ?: throw NullPointerException("Got a null value from the AST"))
        // TODO: parameters might have to be transformed!
        return Expression.GenericExpression(
                parts = parts
        )
    }

    /**
     * Maps a C assignment to a Java assignment.
     */
    private fun assignmentToJavaAssignment(ctx: c_grammarParser.AssignmentContext?): Assignment {
        val lhs: Pair<Type?, String> = computeLeftHandSide(ctx?.lhs()
                ?: throw NullPointerException("Got a null value from the AST"))

        val lhsType = lhs.first
        val lhsVariableName = lhs.second

        val jlhs = LeftHandSide(
                type = lhsType,
                variableName = lhsVariableName
        )

        val jrhs = Expression.SimpleValue(
                value = computeRightHandSide(ctx.rhs())
        )
        return Assignment(
                lhs = jlhs,
                rhs = jrhs,
                comment = null
        )
    }

    /**
     * Maps a left-hand side identifier to a Java identifier: type * name.
     */
    private fun computeLeftHandSide(ctx: c_grammarParser.LhsContext): Pair<Type?, String> {
        return if (ctx.variabledecl() != null) {
            val typeAnnotation = ctx.variabledecl().primitivetype().text // TODO: compute this properly! (Could be a custom type!)
            val variableName = ctx.variabledecl().ID().text

            val cTypesToJavaMap = mapOf(
                    "int" to PrimitiveType.INT
                    // TODO: ...
            )

            val basicType = cTypesToJavaMap[typeAnnotation]
                    ?: throw NoSuchElementException("Couldn't map C type to a Java type!")
            Pair(Type.BasicType(basicType), variableName)
        } else {
            val variableName = ctx.ID().text
            Pair(null, variableName)
        }
    }

    /**
     * Maps a right-hand side to a Java RHS.
     */
    private fun computeRightHandSide(ctx: c_grammarParser.RhsContext): String {
        return ctx.text // TODO: recursively handle nested expressions!
    }

    /**
     * Maps a C for-loop  to a Java for-loop.
     */
    private fun forLoopToJavaForLoop(ctx: c_grammarParser.ForloopContext?): Statement {
        val loopVariable = assignmentToJavaAssignment(
                ctx?.assignment() ?: throw NullPointerException("Got a null value from the AST")
        )

        val loopIncrement = Assignment(
                lhs = JavaIrUtil.lhsWithoutTypeAnnotation(loopVariable.lhs),
                rhs = Expression.ArithmeticExpression(
                        lhs = Expression.SimpleValue(ctx.incrementstatement().ID().text), // TODO: should not be a String but a proper type!
                        rhs = Expression.SimpleValue("1"),
                        arithmeticOperator = Expression.ArithmeticExpression.ArithmeticOperator.ADDITION
                ),
                comment = null
        )

        val loopConditionRv = computeExpr(ctx.condition().expression(1))
        val loopCondition = Expression.Condition(
                lhs = Expression.SimpleValue(loopVariable.lhs.variableName),
                rhs = loopConditionRv,
                conditionalOperator = Expression.Condition.ConditionalOperator.LESSEROREQUALS, // TODO: get the real one from ctx!
                negated = false
        )

        val body: Sequence<Statement> = expressionListToJavaStatements(ctx
                .block()
                .statementlist())
                .asSequence()

        return ForLoop(
                loopVariable = loopVariable,
                loopCondition = loopCondition,
                loopIncrement = loopIncrement,
                body = body,
                comment = null
        )
    }

    /**
     * Maps a list of C expressions to a list of Java statements.
     */
    private fun expressionListToJavaStatements(expressionlist: c_grammarParser.StatementlistContext): List<Statement> {
        return expressionlist
                .statement()
                .asSequence()
                .map { expressionToJavaStatement(it) }
                .toList()
    }

    /**
     * Maps a C if-then-else statement to a Java conditional expression.
     */
    private fun ifthenelseToJavaConditionalExpr(ifthenelse: c_grammarParser.IfthenelseContext?): Statement {
        val condition = computeCondition(ifthenelse?.condition()
                ?: throw NullPointerException("Got a null value from the AST"))
        val body: List<Statement> = expressionListToJavaStatements(ifthenelse.block().statementlist())

        // TODO:
//        if (ctx.elseblock() != null) {
//            val elseBody: List<Statement> = expressionListToJavaStatements(ctx.elseblock().block().expressionlist())
//        }

        return IfThenElse(
                condition = condition,
                thenStatements = body,
                elseStatements = null,
                comment = null
        )
    }

    private fun computeCondition(condition: c_grammarParser.ConditionContext): Expression.Condition {
        val lhs = computeExpr(condition.expression(0))
        val rhs = computeExpr(condition.expression(1))
        val cop = computeConditionalOperator(condition)
        return Expression.Condition(
                lhs = lhs,
                rhs = rhs,
                conditionalOperator = cop,
                negated = false // TODO: this should be computed as part of a Pair analogous to COBOL!
        )
    }

    private fun computeExpr(expression: c_grammarParser.ExpressionContext): Expression {
        return when { // TODO: there should be a sum type String x Int x ... that is used in the SimpleValue constructor!
            expression.ID() != null -> Expression.SimpleValue(
                    value = expression.ID().text
            )

            expression.NUMBER() != null -> Expression.SimpleValue(
                    value = expression.NUMBER().text
            )
            // TODO: ...
            else -> throw Exception("Unrecognized expression type!")
        }
    }

    private fun computeConditionalOperator(condition: c_grammarParser.ConditionContext): Expression.Condition.ConditionalOperator {
        return when {
            condition.comparisonoperator().EQUAL() != null -> Expression.Condition.ConditionalOperator.EQUALS
            condition.comparisonoperator().LESSER() != null -> Expression.Condition.ConditionalOperator.LESSER
            condition.comparisonoperator().GREATER() != null -> Expression.Condition.ConditionalOperator.GREATER
            // TODO: ...
            else -> throw Exception("Unrecognized conditional operator!")
        }
    }

    /**
     * Maps a C include to a Java's import file name.
     */
    private fun computeImportName(ctx: c_grammarParser.ImportheaderContext): String {
        return ctx
                .FILEID()
                .text
                .replace(Regex("\\."), "_") // import without separating dot
                .replace("/", "_")
                .replace("-", "_")
    }

}
